llama_index/node_parser/postprocessors/metadata_extractors.py

```diff
@@ -0,0 +1,393 @@
+"""
+Metadata extractors for nodes. Applied as a post processor to node parsing.
+Currently, only `TextNode` is supported.
+
+Supported metadata:
+Node-level:
+    - `SummaryExtractor`: Summary of each node, and pre and post nodes
+    - `QuestionsAnsweredExtractor`: Questions that the node can answer
+    - `KeywordsExtractor`: Keywords that uniquely identify the node
+Document-level:
+    - `TitleExtractor`: Document title, possible inferred across multiple nodes
+
+Unimplemented (contributions welcome):
+Subsection:
+    - Position of node in subsection hierarchy (and associated subtitles)
+    - Hierarchically organized summary
+
+The prompts used to generate the metadata are specifically aimed to help
+disambiguate the document or subsection from other similar documents or subsections.
+(similar with contrastive learning)
+"""
+
+from abc import ABC, abstractmethod
+import json
+from typing import Any, Callable, List, Optional, Sequence, cast
+from functools import reduce
+
+from llama_index.llm_predictor.base import BaseLLMPredictor, LLMPredictor
+from llama_index.node_parser.interface import NodeParserPostProcessor
+from llama_index.prompts.base import Prompt
+from llama_index.schema import BaseNode, TextNode
+
+
+class MetadataExtractorBase(ABC):
+    is_text_node_only = True
+
+    @abstractmethod
+    def __call__(self, nodes: Sequence[BaseNode]) -> None:
+        """Extracts metadata for a sequence of nodes and mutates the nodes in place.
+
+        Args:
+            nodes (Sequence[Document]): nodes to extract metadata from
+
+        """
+
+
+class MetadataExtractor:
+    """Factory for metadata extractors.
+
+    Example:
+
+    ```
+    post_processor = MetadataExtractor(llm_predictor=llm_predictor)
```

Comment:
mentioned this offline - the main change i was going to suggest is alter the factory pattern to do something similar to `get_response_synthesizer` in `llama_index/response_synthesizers/factory.py`

Reply:
or just have the user use the classes directly. Honestly might be the most straightforward UX to have them load it themselves, e.g.

```python
extractors = [
  KeywordExtractor(llm_predictor),
  QuestionsAnsweredExtractor(llm_predictor),
]
```
etc.

Reply:
Hmm, I see. As mentioned offline, I think the main difference between ResponseSynthesizer and MetadataExtractor is for the latter we chain together many extractors. The builder pattern is nice as well because instead of just an enum, we have extractor specific parameters.

Open to discuss more on the right API.

Reply:
My two cents here is I like Jerry's second suggestion. You can build a pipeline by instantiating in a list like that. Also encourages the Metadata extractors to be more modular (and therefore easy for users to add their own custom versions)

Reply:
Yep agreed that forcing users to deal with the classes can also encourage them to play with them e.g. modify and inspect them, and build their own.